src/types.rs

```diff
@@ -802,8 +832,21 @@ impl CK_ATTRIBUTE {
     }
   }
 
-  pub fn get_date(&self) -> CK_DATE {
-    unsafe { mem::transmute_copy(&*self.pValue) }
+  pub fn get_date(&self) -> Result<CK_DATE, Error> {
+    if self.has_invalid_value() {
+      Err(Error::InvalidInput("Attribute has invalid value"))
+    } else {
+      Ok(unsafe { mem::transmute_copy(&*self.pValue) })
+    }
+  }
+
+  /// Check if the value contained by this attribute is invalid or unavailable.
+  /// This function needs to be called to safely interpret the bytes the attribute contains with
+  /// `from_raw_parts`.
+  pub fn has_invalid_value(&self) -> bool {
```

Comment:
we should review if 'invalid' is the right name here. According to the specs, there are a couple of situations when `CK_UNAVAILABLE_INFORMATION` is used.

1. The attribute is sensitive or unextractable
2. The attribute is invalid (or does not exist, the object does not have this attribute)
3. The `ulValueLen` value which reflects the size of memory to the pointer is actually not large enough to hold all the memory.

maybe naming it `is_unavailable()` is a better more generic name? I'm open to suggestions from you as well.

Reply:
I agree with your three conditions. I took the word invalid from this sentence in the spec:

> The constant CK_UNAVAILABLE_INFORMATION is used in the ulValueLen field to denote an invalid or unavailable value.

(on the section describing `CK_ATTRIBUTE`)

I was thinking of naming it `is_invalid_or_unavailable` but might be a bit long 😋 
The thing is, when you consider the attribute as a result of the `get_attribute_value` function, "unavailable" makes sense, because the attribute searched is really unavailable in the object.
However, if you consider it as an independant object, that can live by itself, you could also say that such attribute is "invalid", as in it does not contain a valid information.

You make the decision on that one, I am fine with both. Code documentation will always help 😃 !

Reply:
I think we can do both! Letting the one call the other. It seems to be done throughout the Rust library. Here is an example in the String package: https://doc.rust-lang.org/std/string/struct.String.html#method.trim_right_matches . `trim_right_matches` and `trim_end_matches` are the same. The one simply calls the other.

So let's rewrite two public functions:
- `is_invalid`
- `is_unavailable`

And let the one simply call the other. I'll let you decide which one calls which. For the private function mentioned above: let's just make sure it uses the short-circuited path.

Reply:
Let's make `is_unavailable` the main one, and let `is_invalid` call `is_unavailable`. Rust docs should explain the behaviour (that they are the same and that the one calls the other).